Add Date Headers to "Scan History" screen

KaungMyat-Min opened the pull request:
### Description
Add date headers to scan history list like the ones in activity log screen. #13992 


### To test:
1. Select a website that has scan feature
2. Tap "scan" on home screen of mobile app
3. Tap "History" at top-right of the screen
4. Verify that the date headers shown as expected

<img src="https://user-images.githubusercontent.com/46663626/107807720-eafe7a00-6d96-11eb-8446-7b424625d7cc.jpg" alt="screenshot_scan_history" width="50%"/>


#### PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add Date Headers to "Scan History" screen

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/common/ViewType.kt:
```diff
@@ -16,5 +16,6 @@ enum class ViewType(val id: Int) {
     PROGRESS(14),
-    BACKUP_RESTORE_BULLET(15)
+    BACKUP_RESTORE_BULLET(15),
+    DETECTED_DATE(16)
```
Minor 🔍: Can we prepend `THREAT_` to `DETECTED_DATE` just like other view types for threats?

KaungMyat-Min replied:
I've changed as requested.

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanListItemState.kt:
```diff
@@ -13,9 +13,14 @@ sealed class ScanListItemState(override val type: ViewType) : JetpackListItemSta
     data class ThreatItemState(
         val threatId: Long,
         val isFixing: Boolean = false,
+        val firstDetectAt: UiString,
```
Minor 🔍: Can we rename `firstDetectAt` to `firstDetectedAt`? Seems like a typo to me.

KaungMyat-Min replied:
Instead of firstDetectedAt, I change to firstDetectedDate to resonate with other parts of the code base.

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/ThreatTestData.kt:
```diff
@@ -50,6 +51,36 @@ object ThreatTestData {
+    )
+
+    // endregion data to test date functions
+
```
Minor 🔍:  Unused data statements can be removed from here.

KaungMyat-Min replied:
I've remove unused codes.

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/history/ScanHistoryListViewModel.kt:
```diff
@@ -76,8 +78,18 @@ class ScanHistoryListViewModel @Inject constructor(
 
-    private fun mapToThreatUiStateList(threatList: List<ThreatModel>) =
-            threatList.map { scanThreatItemBuilder.buildThreatItem(it, this::onItemClicked) }
+    private fun mapToThreatUiStateList(threatList: List<ThreatModel>): List<ScanListItemState> {
+        val uiStates = mutableListOf<ScanListItemState>()
+        threatList.forEach { model ->
+            val currentItem = scanThreatItemBuilder.buildThreatItem(model, this::onItemClicked)
+            val previousItem = uiStates.lastOrNull() as? ThreatItemState
+            if (previousItem == null || currentItem.firstDetectAt != previousItem.firstDetectAt) {
+                uiStates.add(ThreatDateItemState(currentItem.firstDetectAt))
+            }
+            uiStates.add(currentItem)
+        }
+        return uiStates
+    }
```
Question ❓ : How do you feel about splitting `mapToThreatUiStateList` into two parts
- one that actually maps to threat item state
- another that adds date labels?

At first glance, it might not be clear enough that this mapping also adds date headers, wdyt?

KaungMyat-Min replied:
I've separated the adding date header to different function.

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanListItemState.kt:
```diff
@@ -13,9 +13,14 @@ sealed class ScanListItemState(override val type: ViewType) : JetpackListItemSta
         ViewType.THREATS_HEADER
     )
 
+    data class ThreatDateItemState(val text:UiString):ScanListItemState(
+        ViewType.DETECTED_DATE
+    )
```
Suggestion: 1. Add spaces around colon
2. Fit into one line

```suggestion
    data class ThreatDateItemState(val text: UiString) : ScanListItemState(ViewType.DETECTED_DATE)
```

KaungMyat-Min replied:
I've format the code as requested.

ashiagr replied:
`ThreatDateItemState(val text:UiString)` 

Space after colon is still missing 😅. We can check this by using `ktlint` locally, let me know if you need any help with this :).

KaungMyat-Min replied:
Thanks for enlightening me 👍, I am not familiar with `ktlint` 😅. When I check with `ktlint` , there are other warnings in scan package, should I fix them all or fix only the one you mention? 

Example warnings:
`
 68:1 Unexpected indentation (12) (should be 8)
  69:1 Unexpected indentation (12) (should be 8)
  70:1 Unexpected indentation (12) (should be 8)
  71:1 Unexpected indentation (12) (should be 8)
  72:1 Unexpected indentation (16) (should be 12)
  73:1 Unexpected indentation (20) (should be 16)
  74:1 Unexpected indentation (16) (should be 12)
  75:1 Unexpected indentation (20) (should be 16)
  76:1 Unexpected indentation (16) (should be 12)
  77:1 Unexpected indentation (12) (should be 8)
  80:1 Unexpected indentation (12) (should be 8)
  83:1 Unexpected indentation (12) (should be 8)
  84:1 Unexpected indentation (16) (should be 12)
  85:1 Unexpected indentation (12) (should be 8)
  88:1 Unexpected indentation (12) (should be 8)
  89:1 Unexpected indentation (20) (should be 12)
  90:1 Unexpected indentation (24) (should be 16)
  91:1 Unexpected indentation (24) (should be 16)
  92:1 Unexpected indentation (24) (should be 16)
  93:1 Unexpected indentation (24) (should be 16)
  94:1 Unexpected indentation (20) (should be 12)
  104:1 Unexpected indentation (12) (should be 8)
  105:1 Unexpected indentation (16) (should be 12)
  106:1 Unexpected indentation (16) (should be 12)
  107:1 Unexpected indentation (16) (should be 12)
  108:1 Unexpected indentation (12) (should be 8)

`

ashiagr replied:
I don't see these indentation warnings if I run `gradlew` `ktlint` command, there must be some configuration issue or these might just be warnings and not errors, we can figure out and discuss on Slack if you want. I'd suggest to fix only the one related to the mentioned line, I don't think CI will allow to merge this PR otherwise.

ashiagr reviewed (commented):
Thank your working on this KaungMyat Min! 

Functionality works as described, its great you could understand existing code and hook up everything at appropriate places. It's nice that you could add unit tests for this addition 👍.

I just left minor comments/ suggestions/ questions. Let me know if you have questions.

Also can you just include a screenshot of the final screen in the description, will get an approval from the designer as well.

Commit:
Refactor for adding Date Headers to "Scan History" screen
- remove unused test data in ThreatTestData.kt
- change value name to reflect with other components
- splitting mapToThreatUiStateList into two parts
  - one that actually maps to threat item state
  - another that adds date labels

Commit:
Refactor for adding Date Headers to "Scan History" screen
- format the code in ScanListItemState.kt

ashiagr reviewed (commented):
Thanks for all these small improvements! Except of the last comment, I've marked all issues as resolved ✅.

@nbradbury, let us know if you have any additional feedback.

Commit:
update indentation in ScanListItemState.kt

Commit:
remove unused import

ashiagr committed:
Merge branch 'develop' into feature/scanHistoryListDateHeader
